Show users projects in wallet profile [Fixes #73]

wackerow opened the pull request:
### Description
- If currentUser is a recipient, display project info/status in the Profile component (within WalletWidget)
- Adds styling for `isLocked` and `isHidden` cases
- Links to project directly
- (Refactors the `toggleProfile` prop to an emitted 'close' event)

### Screenshot
![image](https://user-images.githubusercontent.com/54227730/131150989-314a8115-5d5a-401f-88fc-cf3c0f457b7c.png)

Open to suggested changes for handling the `isHidden` and `isLocked` states

### Related issue
Fixes #73
https://www.notion.so/efdn/Display-project-thumbnail-title-and-link-in-Profile-7d046a28d5fd43cd8eb0275d65d23915

Fixes: https://github.com/ethereum/clrfund/issues/311

wackerow committed:
adds projects list to user Profile

wackerow committed:
refactor toggleProfile prop to $emit('close')

wackerow committed:
add scrolling for Profile if projects list long

wackerow committed:
object-fit: cover for thumbnails

corwintines committed:
feat: close when clicking Start Contributing in BrightId widget

corwintines committed:
Merge branch 'develop' into wallet-projects

corwintines commented:
@wackerow added the close event to "Start Contributing" in BrightID widget to address https://github.com/ethereum/clrfund/issues/311 as well here. Seemed like the right PR to do so.

wackerow committed:
patch, linting

corwintines commented on vue-app/src/views/Profile.vue:
```diff
@@ -289,6 +343,29 @@ p.no-margin {
+        flex: 1;
+        display: flex;
+        flex-direction: column;
+        justify-content: space-between;
```
This is great. The only thing I was curious about some opinions on is whether we should `justify-conten`t as `space-between` or as `center`. I like how the title of the project is centered this way, but was wondering what some other peoples opinions on this might be. 
<img width="401" alt="Screen Shot 2021-08-29 at 8 07 34 PM" src="https://user-images.githubusercontent.com/15589226/131275649-d24ea6c7-6943-49e7-9b6f-2a79365db68d.png">
vs
<img width="400" alt="Screen Shot 2021-08-29 at 8 07 42 PM" src="https://user-images.githubusercontent.com/15589226/131275663-e3877cbf-b0ee-4eed-bb7f-25e982208bc5.png">

and also 

<img width="400" alt="Screen Shot 2021-08-29 at 8 07 07 PM" src="https://user-images.githubusercontent.com/15589226/131275678-90feb173-12e9-497c-b365-969d7cb85319.png">
vs
<img width="406" alt="Screen Shot 2021-08-29 at 8 06 58 PM" src="https://user-images.githubusercontent.com/15589226/131275691-5735152e-0b45-439b-a55d-87b15bb4e1db.png">

Other than that one question, this looks awesome!

samajammin replied:
Not a strong stance stance here, though I suppose I prefer the `center` alignment.

wackerow replied:
Yeah good question, wasn't sure myself. Went with space-between so if given a list of projects with some under review and some not, that would leave all the titles aligned to the top, rather than switching back and forth.

wackerow replied:
Tested this out... doesn't look that bad at all actually, will switch to `center`

samajammin commented on vue-app/src/views/Profile.vue:
```diff
@@ -143,9 +176,28 @@ export default class NavBar extends Vue {
+      currentRound?.votingDeadline.toSeconds()
+    )
+    const userProjects: Project[] = projects.filter(({ address }) =>
+      isSameAddress(address, currentUser?.walletAddress)
```
So this compares the current user's address to the address the project is using to claim funds. This isn't necessarily the same same address as the one used to submit the application.

Is this what we want? Can we also check against that address used to submit the project? Not sure if we can access the submitters address other than looking through event logs 🤔 

In any case I suppose this is better than nothing 😄

wackerow replied:
We can, would need to update `Project` interface. We look through the event logs to build the list of Projects to begin with...

Currently we call `getProjects` in `recipient-registry-optimistic.ts` which uses our favorite `queryFilter` to filter through and find submission events. These are then run through `decodeProject` and converted into the `Project` type, which is where we lose the `msg.sender` information, which is available as `args._requester`.

I can expand the `Project` interface to also hold a `requester?: string` property, and include this in the check here. 
(added in commit 6b13c00)

samajammin replied:
Nice!

samajammin commented on vue-app/src/views/Profile.vue:
```diff
@@ -69,9 +69,38 @@
+              {{ isLocked ? 'Preview' : 'View' }}
+            </button>
+          </div>
+          <loader v-if="isLoading" />
```
I don't think this loader ever displays because while this data is loading the `v-if="projects.length > 0"` in the parent div always evaluates to false.

samajammin replied:
This makes the interaction a bit awkward as there's always a delay before the info appears:

https://user-images.githubusercontent.com/8097623/131417390-23571561-7a4c-439c-94f0-fa3625df5f9e.mp4

Should we have the loading state appear for all users & display "No projects" or something if the user hasn't submitted any? That would at least avoid this jerky UX.

samajammin replied:
Maybe we always display the "Projects" header, then have an initial loading spinner underneath. If we resolve the query & the user has no projects, we can say "You have submitted any projects" in text underneath the "Projects" header.

samajammin reviewed (commented):
Nice! Left a couple of comments, LMK what you think.

wackerow committed:
adds `requester` property to `Project` interface

Allows showing projects in Profile that a user may have submitted as the requester, even if they are not the beneficiary address.

wackerow committed:
fix loader and fallback message

Keeps "Projects" section header visible always. Shows projects if available. Shows loader while loading, and when done, if no projects, defaults to fallback message and hides loader.

wackerow committed:
`justify-content: center;` for .project-details

samajammin commented on vue-app/src/views/Profile.vue:
```diff
@@ -71,7 +71,39 @@
+            </button>
+          </div>
+        </div>
+        <div v-if="!isLoading && projects.length === 0">
```
This still isn't displaying for me:
https://share.getcloudapp.com/5zuN9bRB

samajammin replied:
Ah, Loader isn't imported. I'll add a commit to fix.

samajammin committed:
Import Loader

samajammin reviewed (approved):
Great work on this!

Added a small commit to import the Loader component - please make sure to check your work 😜
